WhateverGreen/kern_igfx.cpp

```diff
@@ -651,6 +719,574 @@ int IGFX::wrapReadAUX(void *that, IORegistryEntry *framebuffer, uint32_t address
+				DBGLOG("igfx", "SC: LSPCON::getMode() DInfo: [FB%d] The current mode value is 0x%02x.", index, hwModeValue);
+				*mode = parseMode(hwModeValue);
+				break;
+			}
+			
+			// Sleep 1 ms just in case the adapter
+			// is busy processing other I2C requests
+			IOSleep(1);
+		}
+	}
+	
+	return retVal;
+}
+
+IOReturn IGFX::LSPCON::setMode(Mode newMode) {
+	// Guard: The given new mode must be valid
+	if (newMode == Mode::Invalid)
+		return kIOReturnAborted;
+	
+	// Guard: Write the new mode
+	uint8_t hwModeValue = getModeValue(newMode);
+	IOReturn retVal = callbackIGFX->advWriteI2COverAUX(controller, framebuffer, displayPath, DP_DUAL_MODE_ADAPTER_I2C_ADDR, DP_DUAL_MODE_LSPCON_CHANGE_MODE, 1, &hwModeValue, 0);
+	if (retVal != kIOReturnSuccess) {
+		SYSLOG("igfx", "SC: LSPCON::setMode() Error: [FB%d] Failed to set the new adapter mode. RV = 0x%llx.", index, retVal);
+		return retVal;
+	}
+	
+	// Read the register again and verify the mode
+	uint32_t timeout = 200;
+	Mode mode = Mode::Invalid;
+	while (timeout != 0) {
+		retVal = getMode(&mode);
+		// Guard: Read the current effective mode
+		if (retVal != kIOReturnSuccess) {
+			SYSLOG("igfx", "SC: LSPCON::setMode() Error: [FB%d] Failed to read the new effective mode. RV = 0x%llx.", index, retVal);
+			continue;
+		}
+		// Guard: The new mode is effective now
+		if (mode == newMode) {
+			DBGLOG("igfx", "SC: LSPCON::setMode() DInfo: [FB%d] The new mode is now effective.", index);
+			isActive = newMode == Mode::ProtocolConverter;
+			return kIOReturnSuccess;
+		}
+		timeout -= 20;
+		IOSleep(20);
+	}
+	
+	SYSLOG("igfx", "SC: LSPCON::setMode() Error: [FB%d] Timed out while waiting for the new mode to be effective. Last RV = 0x%llx.", index, retVal);
+	return retVal;
+}
+
+IOReturn IGFX::LSPCON::setModeIfNecessary(Mode newMode) {
+	if (isRunningInMode(newMode)) {
+		DBGLOG("igfx", "SC: LSPCON::setModeIfNecessary() DInfo: [FB%d] The adapter is already running in %s mode. No need to update.", index, getModeString(newMode));
+		return kIOReturnSuccess;
+	}
+	
+	return setMode(newMode);
+}
+
+IOReturn IGFX::LSPCON::wakeUpNativeAUX() {
+	uint8_t byte;
+	IOReturn retVal = wrapReadAUX(controller, framebuffer, 0x00000, 1, &byte, displayPath);
+	if (retVal != kIOReturnSuccess)
+		SYSLOG("igfx", "SC: LSPCON::wakeUpNativeAUX() Error: [FB%d] Failed to wake up the native AUX channel. RV = 0x%llx.", index, retVal);
+	else
+		DBGLOG("igfx", "SC: LSPCON::wakeUpNativeAUX() DInfo: [FB%d] The native AUX channel is up. DPCD Rev = 0x%02x.", index, byte);
+	return retVal;
+}
+
+IOReturn IGFX::wrapReadI2COverAUX(void *that, IORegistryEntry *framebuffer, void *displayPath, uint32_t address, uint16_t length, uint8_t *buffer, bool intermediate, uint8_t flags) {
+	if (callbackIGFX->verboseI2C) {
+		uint32_t index = 0xFF;
+		AppleIntelFramebufferExplorer::getIndex(framebuffer, index);
+		SYSLOG("igfx", "SC:  ReadI2COverAUX() called. FB%d: Addr = 0x%02x; Len = %02d; MOT = %d; Flags = %d.",
```

Comment:
Ew, sorry, did not see that previously. I am pretty sure it must be DBGLOG. Here and below, also in wrapWriteI2COverAUX

Reply:
Hmm, `verboseI2C` is enabled by `-igfxi2cdbg`, so if an user has added this boot argument, I think user should always see this line even though he is using the RELEASE version. That's why I choose SYSLOG.

Reply:
Wait, it makes more sense to use the DEBUG version with this boot argument, I will fix it now.

Reply:
Done.